doc/dev/dialects.rst

```diff
@@ -0,0 +1,422 @@
+  Note that while the usage of ODS is strongly recommended, full control over the underlying
+  objects can always be achieved by directly working with the relevant C++ classes.
+
+  In Catalyst, hand-written implementations for classes & methods can be found in the source
+  directory, under `catalyst/mlir/lib\/<DialectName\>/IR <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/lib/Quantum/IR>`_.
+
+An alternative tutorial on creating a basic MLIR dialect can also be found in the official docs, see
+`Creating a Dialect <https://mlir.llvm.org/docs/Tutorials/CreatingADialect/>`_.
+
+
+Creating a new dialect
+======================
+
+Let's see how we can add a new dialect to Catalyst, using OpenQASM as an example! Note that all
+auto-generated files and build artifacts mentioned in the guide will only appear at the end of
+the `Using the dialect <#using-the-dialect>`_ section when the ``make dialects`` command is run.
+
+Start by creating a new TableGen file, located at ``mlir/include/OpenQASM/OpenQASM.td``, with the
+following content:
+
+.. code-block:: tablegen
+
+    include "mlir/IR/DialectBase.td"
+
+    def OpenQASMDialect : Dialect {
+        let summary = "An OpenQASM 3 dialect in MLIR.";
+        let description = [{
+            ...
+        }];
+
+        /// This is the namespace of the dialect in MLIR, which is used as a prefix for types and ops.
+        let name = "oq";
+
+        /// This is the C++ namespace that the dialect, and all sub-components, get placed in.
+        let cppNamespace = "::catalyst::openqasm";
+
+        /// Use the default type printing/parsing hooks, otherwise we would explicitly define them.
+        let useDefaultTypePrinterParser = 1;
+    }
+
+Note the use of ``def`` in TableGen indicates that we are defining a new type of object - under the
+hood this results in a new C++ class. To avoid repeating common parts of object definitions, this
+object creation mechanism itself supports classes. ``def OpenQASMDialect : Dialect`` defines an
+instance of the (built-in) ``Dialect`` TableGen type. In this way, ``OpenQASMDialect`` will inherit
+all properties defined for the ``Dialect`` type.
+
+We could also have subtyped the ``Dialect`` type with ``class QuantumDialect : Dialect``, and then
+created two kinds of quantum dialects as ``def OpenQASMDialect : QuantumDialect`` and
+``def QIRDialect : QuantumDialect`` - if we thought that the two shared some common elements.
+Objects defined with ``def`` are terminal and can no longer be subtyped or inherited from.
+
+Similarly, a new data type for our dialect can be added by via the built-in ``TypeDef`` class
+(in the same file):
+
+.. code-block:: tablegen
+
+    include "mlir/IR/AttrTypeBase.td"
+
+    class OpenQASM_Type<string name, string nameInIR> : TypeDef<OpenQASMDialect, name, []> {
+        let mnemonic = nameInIR;
+    }
+
+    def QubitType : OpenQASM_Type<"Qubit", "qubit"> {
+        let summary = "A single quantum bit reference.";
+    }
+
```

Comment:
It might be worth splitting the code that goes to the header and the implementation code, and explicitly state: "first include these files at the header", then state: "add these code to implement this and that". Users might just copy-paste the code blocks, unaware of the separation between include and implementation sections.

Reply:
What do you mean? I don't think there is any issue mixing `include` commands in-between other tablegen code, if that's what you mean. (the code is meant to be copy-paste-able as is)
And I would much rather have "highly local" code, where you can immediately see why something was included for example, rather than have a big include block at the top that is devoid of context.

Reply:
Makes sense too